presto-spi/src/main/java/com/facebook/presto/spi/LazyIterable.java

```diff
@@ -0,0 +1,20 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi;
+
+public interface LazyIterable<T>
```

Comment:
Thanks for the review and feedback, the entire thinking is described as follows, please let me know if there is anything unreasonable. 

There are several points to consider when implementing setting max threshold for partitions lazy loading:

1. If any delete files are found during the file scanning process, the scanning and loading of partition values would be no longer continue, but instead a single value representing unpartitioned should be returned. So we cannot just scan a part of files to get a specified number of available partition values. So it seems that we cannot set the max threshold and terminate the scan in advance through methods such like `Stream.of(new Iterable<>(...){}).limit(<N>)...`.

2. The max threshold can be passed into `IcebergPartitionLoader` and used there during the loading to terminate the entire scanning process. Because during the loading process, even if no delete files are encountered, when the number of loaded available partitions exceeds the max threshold, we can terminate directly and return a value representing unpartitioned. So we have to figure out a way to pass `max threshold` into `IcebergPartitionLoader`.

3. The construction of the table layout, that is, the lazy partitions and `DiscretePredicates` information is built earlier than the actual loading. Meanwhile, the value of max threshold can only be determined when actual loading is required. For example, in `IcebergMetadataOptimizer`, if further reducible optimization is available, we do not limit the max number of partitions loaded; otherwise, the max number of loaded partitions is set 1000 by default. So it seems that we need to support this delay setting of maximum threshold for the partition loader.

4. `LazyLoadedPartitions` and `PartitionLoader` are both defined in `presto-hive-common`, while `DiscretePredicates` is defined in `presto-spi`, which do not depend on `presto-hive-common`. As mentioned above, we need a way to set the max threshold into `PartitionLoader` through `DiscretePredicates` before actual loading. So we define an interface in `presto-spi` and let `LazyLoadedPartitions` in `presto-hive-common` to implement it, in this way, max threshold can be finally passed into `PartitionLoader` through `LazyLoadedPartitons`.